scipy/_lib/_array_api.py

```diff
@@ -312,6 +312,13 @@ def xp_assert_close(actual, desired, *, rtol=None, atol=0, check_namespace=True,
     __tracebackhide__ = True  # Hide traceback for py.test
     if xp is None:
         xp = array_namespace(actual)
+
+    if (isinstance(actual, (int, float, complex)) and
+        isinstance(desired, (int, float, complex))
+    ):
+        actual = xp.asarray(actual)
+        desired = xp.asarray(desired)
```

Comment:
could you explain why you have added this? Line 259 already includes handling for `xp.isscalar(actual) and xp.isscalar(desired)`.

Reply:
This is to handle python scalars. Otherwise

```
In [2]: k = 5

In [3]: xp_assert_close(k, 5)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[3], line 1
----> 1 xp_assert_close(k, 5)

File ~/repos/scipy/scipy/build-install/lib/python3.12/site-packages/scipy/_lib/_array_api.py:328, in xp_assert_close(actual, desired, rtol, atol, check_namespace, check_dtype, check_shape, check_0d, err_msg, xp)
    323     xp = array_namespace(actual)
    324 desired = _strict_check(actual, desired, xp, check_namespace=check_namespace,
    325                         check_dtype=check_dtype, check_shape=check_shape,
    326                         check_0d=check_0d)
--> 328 floating = xp.isdtype(actual.dtype, ('real floating', 'complex floating'))
    329 if rtol is None and floating:
    330     # multiplier of 4 is used as for `np.float64` this puts the default `rtol`
    331     # roughly half way between sqrt(eps) and the default for
    332     # `numpy.testing.assert_allclose`, 1e-7
    333     rtol = xp.finfo(actual.dtype).eps**0.5 * 4

AttributeError: 'int' object has no attribute 'dtype'

```

Reply:
Thanks, would you be happy with this diff instead?

```diff
diff --git a/scipy/_lib/_array_api.py b/scipy/_lib/_array_api.py
index 2c00095b73..205217622d 100644
--- a/scipy/_lib/_array_api.py
+++ b/scipy/_lib/_array_api.py
@@ -271,7 +271,7 @@ def _strict_check(actual, desired, xp, *,
         assert actual.shape == desired.shape, _msg
 
     desired = xp.broadcast_to(desired, actual.shape)
-    return desired
+    return actual, desired
 
 
 def _assert_matching_namespace(actual, desired):
@@ -312,16 +312,11 @@ def xp_assert_close(actual, desired, *, rtol=None, atol=0, check_namespace=True,
     __tracebackhide__ = True  # Hide traceback for py.test
     if xp is None:
         xp = array_namespace(actual)
-
-    if (isinstance(actual, (int, float, complex)) and
-        isinstance(desired, (int, float, complex))
-    ):
-        actual = xp.asarray(actual)
-        desired = xp.asarray(desired)
-
-    desired = _strict_check(actual, desired, xp, check_namespace=check_namespace,
-                            check_dtype=check_dtype, check_shape=check_shape,
-                            check_0d=check_0d)
+    actual, desired = _strict_check(
+        actual, desired, xp,
+        check_namespace=check_namespace, check_dtype=check_dtype,
+        check_shape=check_shape, check_0d=check_0d
+    )
 
     floating = xp.isdtype(actual.dtype, ('real floating', 'complex floating'))
     if rtol is None and floating:
```

Reply:
Good point, this makes a lot of sense! The lack of symmetry between `actual` and `desired` was looking iffy. Done in https://github.com/scipy/scipy/pull/21512/commits/4f1cc7dbeadd8a6d2cc2f4b91c2b2cd8ba5f0454